std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,576 @@
+    pol commit c[5 * 64];
+    pol commit c_prime[5 * 64];
+    pol commit a_prime[5 * 5 * 64];
+    pol commit a_prime_prime[5 * 5 * 4];
+    pol commit a_prime_prime_0_0_bits[64];
+    pol commit a_prime_prime_prime_0_0_limbs[4];
+
+    // Initially, the first step flag should be 1 while the others should be 0.
+    // builder.when_first_row().assert_one(local.step_flags[0]);
+    // for i in 1..NUM_ROUNDS {
+    //     builder.when_first_row().assert_zero(local.step_flags[i]);
+    // }
+    // for i in 0..NUM_ROUNDS {
+    //     let current_round_flag = local.step_flags[i];
+    //     let next_round_flag = next.step_flags[(i + 1) % NUM_ROUNDS];
+    //     builder
+    //         .when_transition()
+    //         .assert_eq(next_round_flag, current_round_flag);
+    // }
+
+    let step_flags: col[NUM_ROUNDS] = array::new(NUM_ROUNDS, |i| |row| if row % NUM_ROUNDS == i { 1 } else { 0 } );
+
+    // let main = builder.main();
+    // let (local, next) = (main.row_slice(0), main.row_slice(1));
+    // let local: &KeccakCols<AB::Var> = (*local).borrow();
+    // let next: &KeccakCols<AB::Var> = (*next).borrow();
+
+    // let first_step = local.step_flags[0];
+    // let final_step = local.step_flags[NUM_ROUNDS - 1];
+    // let not_final_step = AB::Expr::one() - final_step;
+
+    let first_step: expr = step_flags[0]; // aliasing instead of defining a new fixed column
+    let final_step: expr = step_flags[NUM_ROUNDS - 1];
+    // Equals to 1 for all non-last rows in all 24-row blocks. Forced to zero for the very last row of the submachine (if partial block).
+    col fixed active_row = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]* + [0];
+
+    // // If this is the first step, the input A must match the preimage.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         for limb in 0..U64_LIMBS {
+    //             builder
+    //                 .when(first_step)
+    //                 .assert_eq(local.preimage[y][x][limb], local.a[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    array::zip(preimage, a, |p_i, a_i| first_step * (p_i - a_i) = 0);
+
+    // // The export flag must be 0 or 1.
+    // builder.assert_bool(local.export);
+
+    // force_bool(export);
+
+    // // If this is not the final step, the export flag must be off.
+    // builder
+    //     .when(not_final_step.clone())
+    //     .assert_zero(local.export);
+
+    // not_final_step * export = 0;
+
+    // // If this is not the final step, the local and next preimages must match.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         for limb in 0..U64_LIMBS {
+    //             builder
+    //                 .when(not_final_step.clone())
+    //                 .when_transition()
+    //                 .assert_eq(local.preimage[y][x][limb], next.preimage[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    array::new(100, |i| active_row * (preimage[i] - preimage[i]') = 0);
```

Comment:
Maybe, you could use `unchanged_until(preimage, 1 - first_step);` in here

Reply:
You right!

Reply:
Done